### ISSUE TITLE
Worker lock files are not removed when worker is done with tasks

### ISSUE THREAD

**Issue by hacktobeer** [google/turbinia#905]
The lock files created by the worker in below functionality are never removed, this works as filelock.Filelock is executed in the same worker code to test if a new task can be started but does not work if we want other pieces of code rely on the lock file to see if the worker is busy or not (eg the scaledown functionality)

https://github.com/google/turbinia/blob/master/turbinia/task_manager.py#L107

**Comment by hacktobeer** [comment]
Also, if the actual task executed under the lock crashes the lock will never be released and the worker will not pickup new tasks.

**Comment by hacktobeer** [comment]
See below commit for fix.
https://github.com/google/turbinia/commit/2b2dac3f36a9dc463864b5b09ce885ba24c6cf07

**Comment by aarontp** [comment]
The existence of the lock file isn't intended by the library to be used as evidence of the lock being in place or not.  Other processes or worker code can still use the library interface to check whether it is locked or not.  Can we just have whatever script needs to check this use the intended library interface instead of trying to keep lock counts and delete it ourself?

Regarding crashing workers:  We can just have the worker delete the file on start.

**Comment by hacktobeer** [comment]
The script is a bash one- liner using a preStop hook to check if the worker
is busy or not. How would we use that interface?  I prefer not to create
python scripts to check if a file exists or not. As the library does not
guarantee locking outside of its own thread that would not work anyway.
I understand locking is hard, that's why the library is not deleting the
lock file when the acquire is done.

**Comment by aarontp** [comment]
We could potentially have a couple line python script that checks the lock and is called by the bash one-liner.  It's not ideal, but I think it's probably still better than rolling our own lock counts and file removal, etc, as it would be nice to rely on the library for this as that's what it's there for.  At minimum is there a way to use the libraries internal lock counts?  The PSQ code is threaded, so I'm keen to let the library deal with as much of this as possible rather than have to worry about race conditions and deadlocks ourselves.

Are you sure the library doesn't guarantee locking outside of it's thread?  Isn't that the point of keeping the lock state in a file and not local to the thread?

**Comment by hacktobeer** [comment]
Let me look into this tomorrow.

The code I looked at just drops a file as lock mechanism and handles the
rest in the code scope. So anything outside of that code scope will just
look at the file... which is what we are doing as well.
The thing I am worried about is that when the lock is released by the
library the file is *not* removed (this is by design). So anything outside
of our limited task manager code scope will *only* have a file to look at
as it's not aware of any file lock state. Unless the library keeps system
wide state somewhere of all locks which I doubt as that would also mean
storing the compete context of what's actually being locked.
In short, I'll dig into the internals tomorrow.
Btw, the intern lock count is unfortunately not exposed.

**Comment by hacktobeer** [comment]
Okay, seems the Filelock call uses the flock sys call to keep open file
descriptors to the lock file. This could actually work across
thread/process boundries. Let me run some tests tomorrow with a separate
python scripts,  I have some ideas. Thanks for pointing me in the right
direction!

https://manpages.debian.org/bullseye/manpages-dev/flock.2.en.html

**Comment by wajihyassine** [comment]
Seems like already heading towards that direction, but can maybe check with command like `lsof` if the file is open?

Is this so we can get the grace termination period lower if the worker isn't busy when its terminating? I know current `preStop` command just creates the turbinia-to-delete.lock file immediately (no checks) since pod is set to `TERMINATING` anyways and don't want schedule anymore work on it regardless if something is running or not

**Comment by hacktobeer** [comment]
Will check lsof, good one.

This is indeed to facilitate faster cleanup of workers that are idle
instead of having them always running  until maximum timeout is reached.

**Comment by hacktobeer** [comment]
Fix in turbinia-ng branch -> https://github.com/google/turbinia/commit/2b2dac3f36a9dc463864b5b09ce885ba24c6cf07#diff-da83cfaca362f241b9f852f9177e920ff33872da6385bceec7860fb00bf3945e

Decide on merging later.
